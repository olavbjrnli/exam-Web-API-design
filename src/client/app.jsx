import React from "react";
import {BrowserRouter, Switch, Route} from 'react-router-dom'

import {Home} from "./home";
import Login from "./login";
import SignUp from "./signup";
import Details from "./details";

//Taken and very inspired by https://github.com/arcuri82/web_development_and_api_design/blob/master/les08/authentication/src/client/index.jsx
export class App extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            userId: null
        };
    }

    updateLoggedInUserId = (userId) => {
        this.setState({userId: userId});
    };

    notFound() {

        return (
            <div>
                <h2>NOT FOUND: 404</h2>
                <p>
                    Ups: This page is either not available or doesn't exist.
                </p>
            </div>
        );

    };

    render() {


        return (
            <BrowserRouter>
                <div>
                    <Switch>
                        <Route exact path="/login"
                               render={props => <Login {...props}
                                                       userId={this.state.userId}
                                                       updateLoggedInUserId={this.updateLoggedInUserId}/>}/>
                        <Route exact path="/signup"
                               render={props => <SignUp {...props}
                                                        userId={this.state.userId}
                                                        updateLoggedInUserId={this.updateLoggedInUserId}/>}/>
                        <Route exact path="/"
                               render={props => <Home {...props}
                                                      userId={this.state.userId}
                                                      updateLoggedInUserId={this.updateLoggedInUserId}/>}/>
                        <Route exact path="/details"
                               render={props => <Details {...props}
                                                         userId={this.state.userId}
                                                         updateLoggedInUserId={this.updateLoggedInUserId}/>}/>
                        <Route component={this.notFound}/>


                    </Switch>
                </div>
            </BrowserRouter>
        );
    }
}

