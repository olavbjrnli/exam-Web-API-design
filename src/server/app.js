const express = require('express');
const bodyParser = require('body-parser');
const passport = require('passport');
const session = require("express-session");
const LocalStrategy = require('passport-local').Strategy;
const path = require('path');
const cors = require('cors');
const routes = require('./routes');
const Repository = require('./repository');

//Created with https://github.com/arcuri82/web_development_and_api_design/blob/master/les08/authentication/src/server/app.js
const app = express();


if(false){
    console.log("Using CORS to allow all origins");
    app.use(cors());

}

//to handle JSON payloads
app.use(bodyParser.json());

//to handle Form POST. "extended" is just to be able to parse all kinds of objects
app.use(bodyParser.urlencoded({extended: true}));


app.use(session({
    secret: 'a secret used to encrypt the session cookies',
    resave: false,
    saveUninitialized: false
}));


//needed to server static files, like HTML, CSS and JS.
app.use(express.static('public'));


passport.use(new LocalStrategy(

    {
        usernameField: 'userId',
        passwordField: 'password'
    },
    function (userId, password, done) {

        const ok = Repository.verifyUser(userId, password);

        if (!ok) {
            return done(null, false, {message: 'Invalid username/password'});
        }

        const user = Repository.getUser(userId);
        return done(null, user);
    }
));


passport.serializeUser(function (user, done) {
    done(null, user.id);
});

passport.deserializeUser(function (id, done) {

    const user = Repository.getUser(id);

    if (user !== undefined) {
        done(null, user);
    } else {
        done(null, false);
    }
});

app.use(passport.initialize());
app.use(passport.session());


//--- Routes -----------
app.use('/', routes);

//handling 404
app.use((req, res, next) => {
    res.sendFile(path.resolve(__dirname, '..', '..', 'public', 'index.html'));
});

module.exports = app;
