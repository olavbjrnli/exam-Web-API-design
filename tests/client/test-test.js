import * as ReactDOM from "enzyme";
import {App} from "../../src/client/app.jsx";
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {shallow} from "enzyme";
import Details from "../../src/client/details.jsx";
import Pokemonlist from "../../src/client/PokemonList";
import Pokemon from "../../src/client/pokemon";
configure({ adapter: new Adapter() });

const React = require('react');
const app = require('../../src/server/app');
const request = require('supertest');

// taken from https://stackoverflow.com/questions/41797169/basic-react-tests-failing-after-implementing-redux
it('App renders without crashing', () => {
    const div = document.createElement('div'); // create the div here
    ReactDOM.render(<App />, div);
});
it('Details renders without crashing', () => {
    const div = document.createElement('div'); // create the div here
    ReactDOM.render(<Details />, div);

});
it('Pokemonlist renders without crashing', () => {
    const div = document.createElement('div'); // create the div here
    ReactDOM.render(<Pokemonlist />, div);

});
it('Pokemon renders without crashing', () => {
    const div = document.createElement('div'); // create the div here
    ReactDOM.render(<Pokemon />, div);

});


//https://medium.com/codeclan/testing-react-with-jest-and-enzyme-20505fec4675
describe('App', () => {
    it('should render correctly in "debug" mode', () => {
        const component = shallow(<App debug />);

        expect(component).toMatchSnapshot();
    });
});

