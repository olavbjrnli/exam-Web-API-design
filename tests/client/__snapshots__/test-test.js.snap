// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly in "debug" mode 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    debug={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <div>
        <Switch>
          <Route
            exact={true}
            path="/login"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/signup"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/details"
            render={[Function]}
          />
          <Route
            component={[Function]}
          />
        </Switch>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Switch>
          <Route
            exact={true}
            path="/login"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/signup"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/details"
            render={[Function]}
          />
          <Route
            component={[Function]}
          />
        </Switch>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Route
              exact={true}
              path="/login"
              render={[Function]}
            />,
            <Route
              exact={true}
              path="/signup"
              render={[Function]}
            />,
            <Route
              exact={true}
              path="/"
              render={[Function]}
            />,
            <Route
              exact={true}
              path="/details"
              render={[Function]}
            />,
            <Route
              component={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "exact": true,
              "path": "/login",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "exact": true,
              "path": "/signup",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "exact": true,
              "path": "/",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "exact": true,
              "path": "/details",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <div>
          <Switch>
            <Route
              exact={true}
              path="/login"
              render={[Function]}
            />
            <Route
              exact={true}
              path="/signup"
              render={[Function]}
            />
            <Route
              exact={true}
              path="/"
              render={[Function]}
            />
            <Route
              exact={true}
              path="/details"
              render={[Function]}
            />
            <Route
              component={[Function]}
            />
          </Switch>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Switch>
            <Route
              exact={true}
              path="/login"
              render={[Function]}
            />
            <Route
              exact={true}
              path="/signup"
              render={[Function]}
            />
            <Route
              exact={true}
              path="/"
              render={[Function]}
            />
            <Route
              exact={true}
              path="/details"
              render={[Function]}
            />
            <Route
              component={[Function]}
            />
          </Switch>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Route
                exact={true}
                path="/login"
                render={[Function]}
              />,
              <Route
                exact={true}
                path="/signup"
                render={[Function]}
              />,
              <Route
                exact={true}
                path="/"
                render={[Function]}
              />,
              <Route
                exact={true}
                path="/details"
                render={[Function]}
              />,
              <Route
                component={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "exact": true,
                "path": "/login",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "exact": true,
                "path": "/signup",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "exact": true,
                "path": "/",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "exact": true,
                "path": "/details",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
